### PR TITLE
[urijs] Fix argument type for search/query

### PR DESCRIPTION
**Pull request by PikachuEXE**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://medialize.github.io/URI.js/docs.html#accessors-search
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

My last PR #47099 got incorrect test case thus unable to cover the type that should really be accepted

### CODE REVIEW

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## urijs ([<kbd>unpkg</kbd>](https://unpkg.com/browse/urijs@latest/))
was missing the following properties:
1. version
1. protocol_expression
1. idn_expression
1. punycode_expression
1. ip4_expression

<details>
<summary>as well as these 19 other properties...</summary>
<p>ip6_expression, find_uri_expression, findUri, defaultPorts, hostProtocols, invalid_hostname_characters, domAttributes, getDomAttribute, characters, decodePath, decodeUrnPath, recodePath, recodeUrnPath, buildQueryParameter, setQuery, hasQuery, ensureValidHostname, ensureValidPort, noConflict</p>
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against bcceffd0e97f339a04f62cdcfcca489faffdeed3
</p>

File: types/urijs/index.d.ts [hunk 1]
```diff
@@ -212,7 +212,8 @@ interface URI {
     preventInvalidHostname(val: boolean): URI;
 
     query(): string;
-    query(qry: string | URI.QueryDataMap | ((qryObject: URI.QueryDataMap) => (URI.QueryDataMap | undefined))): URI;
+    // tslint:disable-next-line void-return
+    query(qry: string | URI.QueryDataMap | ((qryObject: URI.QueryDataMap) => (URI.QueryDataMap | void))): URI;
```
Comment (marcelltoth) [inline types/urijs/index.d.ts:216]: While this next version is more correct, returning undefined also works (has the same effect as `void`). So I'd say let's keep both.

What do you think?

Comment (marcelltoth) [reply types/urijs/index.d.ts:216]: image

File: types/urijs/index.d.ts [hunk 2]
```diff
@@ -229,7 +230,8 @@ interface URI {
     scheme(): string;
     scheme(protocol: string): URI;
     search(): string;
-    search(qry: string | URI.QueryDataMap | ((qryObject: URI.QueryDataMap) => (URI.QueryDataMap | undefined))): URI;
+    // tslint:disable-next-line void-return
+    search(qry: string | URI.QueryDataMap | ((qryObject: URI.QueryDataMap) => (URI.QueryDataMap | void))): URI;
```
Comment (marcelltoth) [inline types/urijs/index.d.ts:234]: Ditto

**Review by marcelltoth — CHANGES_REQUESTED** [review]

File: types/urijs/index.d.ts (hunk 2 above)
Comment (XVincentX) [reply types/urijs/index.d.ts:234]: I'd disable such rule `// tslint:disable-next-line void-return` directly in the file. If it's the way the library works, we do not have to circumnavigate anything.

**Comment by PikachuEXE** [comment]
@marcelltoth 
`URI.QueryDataMap | void` actually accepts function with return value of `undefined`
But `URI.QueryDataMap | undefined` does not accept function with no return 
(Error message: `Type 'void' is not assignable to type 'QueryDataMap | undefined'`)
image


I have updated the PR to keep the tests like `u.query(() => undefined)`

**Comment by marcelltoth** [comment]
> 
> 
> @marcelltoth
> `URI.QueryDataMap | void` actually accepts function with return value of `undefined`
> But `URI.QueryDataMap | undefined` does not accept function with no return
> (Error message: `Type 'void' is not assignable to type 'QueryDataMap | undefined'`)
> image
> 
> I have updated the PR to keep the tests like `u.query(() => undefined)`

I always learn something. 😉 Looks good then. Consider @XVincentX's suggestion but OK on my side.

**Review by marcelltoth — APPROVED** [review]

**Comment by PikachuEXE** [comment]
@XVincentX 
Without disabling the rule it will disallow `(qryObject: URI.QueryDataMap) => (URI.QueryDataMap | void)` as "it's not return type"
Would be good to know other possible ways to allow functions without return value without using `void`
But to solve annoying type error I am merging it now first

**Comment by PikachuEXE** [comment]
Ready to merge

**Pull request merged**